src/smtp-lib/smtp_server.cpp

```diff
@@ -1,37 +1,61 @@
 #include <thread>
 
 #include "smtp_server.hpp"
-#include "socket.hpp"
+#include "socket_listener.hpp"
 
-SMTPServer::SMTPServer(uint16_t port) : m_port(port), m_is_running(false), m_stop_requested(false) {}
+SMTPServer::SMTPServer(uint16_t port) : m_socket_listener(port), m_is_running(false), m_stop_requested(false) {}
 
 void SMTPServer::run() {
-    m_receiverThread = std::thread(&SMTPServer::accept_connections, this);
+    m_stop_requested = false;
     m_is_running = true;
+
+    accept_connections();
+
+    stop_mail_receivers();
+
+    RawSocket::clean_up();
 }
 
 bool SMTPServer::is_running() { return m_is_running; }
 
 void SMTPServer::stop() {
     m_stop_requested = true;
-    m_receiverThread.join();
-    m_stop_requested = false;
     m_is_running = false;
 }
 
 void SMTPServer::accept_connections() {
-    // This can throw if socket bind/listen fails, but the executable should take care of this
-    Socket socket{m_port};
+    while (!m_stop_requested) {
+        auto connection = accept_connection();
+        if (connection.is_valid()) {
+            add_new_mail_receiver_for(std::move(connection));
+            start_worker_for_last_mail_receiver();
+        }
+    }
+}
 
-    // Start the the handler for a new SMTP request
-    auto handle_client = [](MailReceiver handler) { handler.run(); };
+Connection SMTPServer::accept_connection() { return Connection(m_socket_listener.accept_connection()); }
 
-    // This is the server's main loop to accept connections
-    while (!m_stop_requested) {
-        // This blocks until a new connection comes in.
-        auto smtp_handler = socket.accept_connection();
+void SMTPServer::add_new_mail_receiver_for(Connection connection) {
+    m_mail_receivers.emplace_back(std::move(connection));
```

Comment:
This will leak because we cannot clean up the thread objects that have finished but they will always be in the vector. Any idea how to fix this?

Reply:
This Merge Request aims for getting the old status back without blocking network calls. I have an answer to this question. But I would not implement it here.